### PR TITLE
[Fix] `prop-types`, `propTypes`: add support for exported type inference

### LINKED ISSUE #3015: forwardRef example has error in TypeScript
```
    "@types/react": "^17.0.11",
    "@types/react-dom": "^17.0.8",
    "@typescript-eslint/eslint-plugin": "^4.28.0",
    "@typescript-eslint/parser": "^4.28.0",
```

The code from https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/forward_and_create_ref/ on forwardRef has the error `'children' is missing in props validation` based on the `react/propTypes` ESLint rule https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prop-types.md

This rule says it validates on prop-types typing, as well as TS types. I am not sure if this is an issue in ESLint or TypeScript

### PR DESCRIPTION
**Pull request by vedadeepta**
Fixes #3015

### CODE REVIEW

File: lib/util/propTypes.js [hunk 1]
```diff
@@ -547,6 +547,35 @@ module.exports = function propTypesInstructions(context, components, utils) {
...
+   * @param {ASTNode} node
+   * @param {string} typeName
+   * @return {boolean}
+   */
+  function filterInterfaceOrAliasByName(node, typeName) {
+    if (node.id && node.id.name) {
+      return node.id.name === typeName;
+    }
+
+    if (node && node.declaration && node.declaration.id && node.declaration.id.name) {
```
Comment (ljharb) [inline lib/util/propTypes.js:572]: if `node` is falsy then this function would have crashed on line 568

File: lib/util/propTypes.js [hunk 2]
```diff
@@ -547,6 +547,35 @@ module.exports = function propTypesInstructions(context, components, utils) {
...
+      return node.declaration.id.name === typeName;
+    }
+
+    return false;
```
Comment (ljharb) [inline lib/util/propTypes.js:576]:
```suggestion
    return node && (
      (node.id && node.id.name === typeName)
      || (node.declaration && node.declaration.id && node.declaration.id.name === typeName)
    );
```
?

**Review by ljharb — APPROVED** [review]
LGTM!

File: lib/util/propTypes.js (hunk 1 above)
Comment (vedadeepta) [reply lib/util/propTypes.js:572]: yes, my bad - the check should not contain `node && ` part. `node` would never be falsy.

File: lib/util/propTypes.js (hunk 2 above)
Comment (vedadeepta) [reply lib/util/propTypes.js:576]: sure, but i feel sometimes one liners are less readable.

Comment (ljharb) [reply lib/util/propTypes.js]: yeah true, in this case it'd be much more readable with optional chaining, but we can't use that yet :-/

**Pull request merged**